src/typecheck/mod.rs

```diff
@@ -1042,23 +1042,30 @@ pub fn apparent_type(
 
 /// Infer the type of a non annotated record by gathering the apparent type of the fields. It's
 /// currently used essentially to type the stdlib.
-pub fn infer_type(t: &Term) -> TypeWrapper {
+pub fn infer_record_type(t: &Term) -> TypeWrapper {
     match t {
+        // An explicit annotation must take precedence over this inference, so let's use it.
+        t @ Term::MetaValue(MetaValue {
+            types, contracts, ..
+        }) if (types.is_some() || !contracts.is_empty()) => apparent_type(t, None, None).into(),
+        // We unwrap meta-values without any type information. Otherwise, something like a
+        // top-level module documentation would stop this function from doing its job.
+        Term::MetaValue(MetaValue {
+            value: Some(rt),
+            types: None,
+            contracts,
+            ..
+        }) if contracts.is_empty() => infer_record_type(rt.as_ref()),
```

Comment:
Shouldn't this be `if types.is_none() && contracts.is_empty()` instead?

Reply:
Ah, good catch!

Reply:
In fact, it's already covered by the pattern `types: None`, because types is just an option.